spark3/src/main/java/org/apache/iceberg/spark/Spark3Util.java

```diff
@@ -767,19 +772,46 @@ public static CatalogAndIdentifier catalogAndIdentifier(String description, Spar
     }
   }
 
+  public static CatalogAndIdentifier catalogAndIdentifier(String description, SparkSession spark, String name,
```

Comment:
Rather than making this resolution more complicated with cases for timestamp and snapshot ID, can we make these orthogonal? What about calling `catalogAndIdentifier` and then adding the snapshot to the path identifier?

Reply:
I tried leaving the `catalogAndIdentifier` methods in `Spark3Util` unchanged, and adding the snapshot id and timestamp to the Identifier in (spark3) `IcebergSource#catalogAndIdentifier` after calling `Spark3Util.catalogAndIdentifier`, but I must have missed something because the unit tests I added then failed. So I reverted that experiment. This implementation may be slightly more complicated but I know that it works.

Reply:
I figured out what I missed the first time around. I now add no new methods to `Spark3Util` and use existing `Spark3Util.catalogAndIdentifier` methods to obtain a `CatalogAndIdentifier`, and then swap out the `Identifier` for a snapshot-aware one.